src/lib/operators/insert.cpp

```diff
@@ -107,136 +93,132 @@ std::shared_ptr<const Table> Insert::_on_execute(std::shared_ptr<TransactionCont
-    }
+      // If the last Chunk of the target Table is either immutable or full, append a new mutable Chunk
+      if (!target_chunk->is_mutable() || target_chunk->size() == _target_table->max_chunk_size()) {
+        _target_table->append_mutable_chunk();
+        ++target_chunk_id;
+        target_chunk = _target_table->get_chunk(target_chunk_id);
+      }
 
-    auto remaining_rows = total_rows_to_insert;
-    while (remaining_rows > 0) {
-      auto current_chunk = _target_table->get_chunk(static_cast<ChunkID>(_target_table->chunk_count() - 1));
-      auto rows_to_insert_this_loop = std::min(_target_table->max_chunk_size() - current_chunk->size(), remaining_rows);
+      const auto target_chunk_num_inserted_rows =
+          std::min<size_t>(_target_table->max_chunk_size() - target_chunk->size(), remaining_rows);
 
-      // Resize MVCC vectors.
-      current_chunk->get_scoped_mvcc_data_lock()->grow_by(rows_to_insert_this_loop, MvccData::MAX_COMMIT_ID);
+      _target_chunk_ranges.emplace_back(
+          ChunkRange{target_chunk_id, target_chunk->size(),
+                     static_cast<ChunkOffset>(target_chunk->size() + target_chunk_num_inserted_rows)});
 
-      // Resize current chunk to full size.
-      auto old_size = current_chunk->size();
-      for (ColumnID column_id{0}; column_id < current_chunk->column_count(); ++column_id) {
-        typed_segment_processors[column_id]->resize_vector(current_chunk->get_segment(column_id),
-                                                           old_size + rows_to_insert_this_loop);
-      }
+      // Grow MVCC columns while simultaneously locking them for the current transaction.
+      target_chunk->get_scoped_mvcc_data_lock()->grow_by(target_chunk_num_inserted_rows, context->transaction_id());
 
-      remaining_rows -= rows_to_insert_this_loop;
+      // Grow data Segments.
+      auto old_size = target_chunk->size();
+      for (ColumnID column_id{0}; column_id < target_chunk->column_count(); ++column_id) {
+        resolve_data_type(_target_table->column_data_type(column_id), [&](const auto data_type_t) {
+          using ColumnDataType = typename decltype(data_type_t)::type;
 
-      // Create new chunk if necessary.
-      if (remaining_rows > 0) {
-        _target_table->append_mutable_chunk();
-        end_chunk_id++;
+          resize_value_segment<ColumnDataType>(target_chunk->get_segment(column_id),
+                                               old_size + target_chunk_num_inserted_rows);
+        });
       }
+
+      remaining_rows -= target_chunk_num_inserted_rows;
     }
   }
-  // TODO(all): make compress chunk thread-safe; if it gets called here by another thread, things will likely break.
 
-  // Then, actually insert the data.
-  auto input_offset = 0u;
-  auto source_chunk_id = ChunkID{0};
-  auto source_chunk_start_index = 0u;
+  /**
+   * 2. Insert the Data into the memory allocated in the first step. Write the transaction_context's transaction_id into
```

Comment:
Add a comment that the lock has been released?

Reply:
done